Sources_v3/Workers/Background/MessageEditor.swift

```diff
@@ -57,58 +54,43 @@ class MessageEditor<ExtraData: ExtraDataTypes>: Worker {
         let wasEmpty = pendingMessageIDs.isEmpty
         changes.pendingEditMessageIDs.forEach { pendingMessageIDs.insert($0) }
         if wasEmpty {
-            database.backgroundReadOnlyContext.perform { [weak self] in
-                self?.processNextMessage()
-            }
+            processNextMessage()
         }
     }
 
     private func processNextMessage() {
         guard let messageId = pendingMessageIDs.first else { return }
         
-        guard let dto = database.backgroundReadOnlyContext.message(id: messageId) else {
-            removeMessageIDAndContinue(messageId)
-            return
-        }
-        
-        guard dto.localMessageState == .pendingSync else {
-            removeMessageIDAndContinue(messageId)
-            return
-        }
-        
-        markMessage(withID: messageId, as: .syncing) {
+        backgroundContext.perform { [weak self] in
+            guard let dto = self?.backgroundContext.message(id: messageId), dto.localMessageState == .pendingSync else {
+                self?.removeMessageIDAndContinue(messageId)
+                return
+            }
+            
             let requestBody = dto.asRequestBody() as MessageRequestBody<ExtraData>
-            self.apiClient.request(endpoint: .editMessage(payload: requestBody)) {
-                switch $0 {
-                case .success:
-                    self.markMessage(withID: messageId, as: nil) {
-                        self.removeMessageIDAndContinue(messageId)
-                    }
-                case .failure:
-                    self.markMessage(withID: messageId, as: .syncingFailed) {
-                        self.removeMessageIDAndContinue(messageId)
-                    }
+            
+            self?.markMessage(withID: messageId, as: .syncing)
+            self?.apiClient.request(endpoint: .editMessage(payload: requestBody)) { result in
+                self?.backgroundContext.perform {
+                    self?.markMessage(withID: messageId, as: result.error == nil ? nil : .syncingFailed)
```

Comment:
`markMessage` calls `database.write` (so switches queues) and `removeMessageIDAndContinue` calls `processNextMessage` which in turn calls `backgroundContext.perform`, so do these 2 should be called in another `backgroundContext.perform`?

Reply:
`markMessage` doesn't access `database.write` but work with a `backgroundContext` directly so we don't switch the queues.

Regarding the safety of nested `perform()` calls, I think it's safe as it should just schedule the block execution on the context's private queue (but it's not mentioned directly in the docs as it is mentioned for `performAndWait`: **May safely be called reentrantly**) 😄